### ISSUE TITLE
Bug: [no-shadow] false positive with augmented modules

### ISSUE THREAD

**Issue by dereli** [typescript-eslint/typescript-eslint#5681]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.3&sourceType=module&code=JYWwDg9gTgLgBDAnmApnA3nAsogKstAXzgDMoIQ4ByAOgHolUBnKgbgCh2ATFAYwBsAhlDQgIXAK780tBgRYZ2cBAWx5VAXgykIEAFxwmMKMAB2AcziEOhTigAekWHF4RTRuACNhBnPlRwWpgkugZUAIwATFRWHEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uQAsBDACYcA7uiiJo0DtEjgwAXxDygA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
import type { MyType } from './types';

declare module './types' {
  type MyType = { foo: string };
}

export const bar: MyType = { foo: '' };
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-shadow": "error"
  },
};
```

### Expected Result

No error is expected on line 4 column 8.

### Actual Result

Error is reported: 'MyType' is already declared in the upper scope on line 1 column 15.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.33.0` |
| `@typescript-eslint/parser`            | `5.35.1` |
| `TypeScript`                           | `4.7.4` |
| `ESLint`                               | `8.23.0` |
| `node`                                 | `16.14.2` |

**Comment by JoshuaKGoldberg** [comment]
But... this is a shadowing, no? The rule's intent is to stop it being unclear in inner scopes whether types & variables refer to something in the inner or outer scope. That's what's happening with `MyType` in the example, right?

Do you have a more concrete example that can demonstrate why the rule should be changed?

**Comment by bradzacher** [comment]
It's not *technically* shadowing - it's declaration merging.
This particular form is a really rare pattern, but it's valid.

**Comment by dereli** [comment]
As @bradzacher said this is not shadowing; the import on line 1 is actually importing the newly created type on line 4 and that new type is merged with the types in `./types`.

I have an actual use case which might count as a concrete example. [Knex](https://github.com/knex/knex) has an empty [`Tables` interface](https://github.com/knex/knex/blob/f7ccde80c91fe661696e0c5a14f5e019d90cd453/types/tables.d.ts) which is supposed to be augmented with actual data entities ([documentation here](https://knexjs.org/guide/#typescript)).

Having the data access functions together with the entity types makes a lot of sense and is quite practical in my case:


```js
// filename: account.ts

import type { Account } from 'knex/types/tables';
import { createOne } from '../data-access';

declare module 'knex/types/tables' {
  type Account = {
    id: number;
    // ...
  };
}

export const createAccount = createOne<Account>('accounts');

```

**Comment by JoshuaKGoldberg** [comment]
Got it, thanks. Agreed this is a bug that should be fixed.

I would also think this is an odd and confusing behavior. Although it's not _technically_ shadowing as bradzacher said, it still suffers from some of the same cognitive problems as shadowing. I'll file a new issue. Cheers!

**Comment by thw0rted** [comment]
Hello from the future!  Did this issue ever get resolved?  It's my understanding that module augmentation for interface merging is a pretty common practice, especially when describing the types of existing libraries that use a mixin / plugin type architecture.  I guess we can always disable the rule on a case by case basis with a comment directive, but does it really make sense to apply it to interface merging at all?

**Comment by thw0rted** [comment]
Hah!  Thanks, but what I meant was that your previous comment sounded like the rule proposal (which got closed from lack of interest) was tied to how this issue would be resolved.  It sounds like I misunderstood that.  So, the team's perspective is that interface augmentation should be permitted, as in the example in the OP, and the current flagging behavior is a FP.

I have to admit, I don't know if I'll have the spare cycles to learn enough eslint internals to get a PR together, but it's tempting....

**Comment by JoshuaKGoldberg** [comment]
Temptation is great 😄. If it helps for you / any future readers, https://typescript-eslint.io/blog/asts-and-typescript-eslint -> https://sourcegraph.com/notebooks/Tm90ZWJvb2s6MTA2OA== is a great reading journey to understand this.
